src/nfa/intersection.cc

```diff
@@ -16,177 +16,254 @@
 // MATA headers
 #include "mata/nfa/nfa.hh"
 #include "mata/nfa/algorithms.hh"
+#include <cassert>
+#include <functional>
+
 
 using namespace mata::nfa;
 
 namespace {
 
+using ProductMap = std::unordered_map<std::pair<State,State>,State>;
+using ProductMatrix = std::vector<std::vector<State>>;
+using ProductVecMap = std::vector<std::unordered_map<State,State>>;
+//Unordered map seems to be faster than ordered map here, but still very much slower than matrix.
+
+} // Anonymous namespace.
+
+namespace mata::nfa {
+
+Nfa intersection(const Nfa& lhs, const Nfa& rhs, const Symbol first_epsilon, ProductMap *prod_map) {
+
+    auto both_final = [&lhs,&rhs](const State lhs_state,const State rhs_state) {
+        return lhs.final.contains(lhs_state) && rhs.final.contains(rhs_state);
+    };
+
+    return algorithms::product(lhs, rhs, both_final, first_epsilon, prod_map);
+}
+
+//TODO: move this method to nfa.hh? It is something one might want to use (e.g. for union, inclusion, equivalence of DFAs).
+Nfa mata::nfa::algorithms::product(
+        const Nfa& lhs, const Nfa& rhs, const std::function<bool(State,State)>&& final_condition,
+        const Symbol first_epsilon, ProductMap *product_map) {
+    Nfa product{}; // Product of the intersection.
+    // Product map for the generated intersection mapping original state pairs to new product states.
+    std::deque<State> pairs_to_process{}; // Set of state pairs of original states to process.
+
+    //The largest matrix (product_matrix) of pairs of states we are brave enough to allocate.
+    // Let's way we are fine with allocating large_product * (about 8 Bytes) space.
+    // So ten million cells is close to 100 MB.
+    // If the number is larger, then we do not allocate a matrix, but use a vector of unordered maps (product_vec_map),
+    // together with a heuristic that speeds up containment test, because unordered map lookup is not cheap:
+    // We remember for every lhs/rhs state a range of rhs/lhs states that have already appeared with it (vectors min_lhs/rhs and max_lhs/rhs),
+    // And every containment test first asks whether lhs and rhs are in each others ranges.
+    // This is several times faster compared to pure product_vec_map, which is turn is notably faster that one unordered map from pairs to states.
+    // (but Juraj says that rewriting hash function may help unordered map significantly, so maybe that would be enough ... ?)
+    // TODO: where to put this magical constant? It should not be here.
```

Comment:
I tried, replacing unordered_map with map makes it 40% slower.
StackOverflow people say that this is because memory locality is by far the most important thing when the collections are not too large, and set has pointers everywhere.

Besides, when trying this, I also tried disable the optimization with ranges, and the fucker got faster. 
So maybe we can simplify this and remove all that, just keep the switching between matrix and vector of unordered maps. But I will try to see what happened first.